util/claims/claims.go

```diff
@@ -1,56 +1,33 @@
 package claims
 
 import (
-	"encoding/json"
-
 	"github.com/golang-jwt/jwt/v5"
+
+	jwtutil "github.com/argoproj/argo-cd/v3/util/jwt"
 )
 
-// ArgoClaims defines the claims structure based on Dex's documented claims
-type ArgoClaims struct {
-	jwt.RegisteredClaims
-	Email         string   `json:"email,omitempty"`
-	EmailVerified bool     `json:"email_verified,omitempty"`
-	Name          string   `json:"name,omitempty"`
-	Groups        []string `json:"groups,omitempty"`
-	// As per Dex docs, federated_claims has a specific structure
-	FederatedClaims *FederatedClaims `json:"federated_claims,omitempty"`
-}
+// GetUserIdentifier returns a consistent user identifier, checking federated_claims.user_id when Dex is in use
+func GetUserIdentifier(c jwt.MapClaims) string {
+	if c == nil {
+		return ""
+	}
 
-// FederatedClaims represents the structure documented by Dex
-type FederatedClaims struct {
-	ConnectorID string `json:"connector_id"`
-	UserID      string `json:"user_id"`
-}
+	// Fallback to sub if federated_claims.user_id is not set.
+	fallback := jwtutil.StringField(c, "sub")
 
-// MapClaimsToArgoClaims converts a jwt.MapClaims to a ArgoClaims
-func MapClaimsToArgoClaims(claims jwt.MapClaims) (*ArgoClaims, error) {
-	if claims == nil {
-		return &ArgoClaims{}, nil
+	f := c["federated_claims"]
+	if f == nil {
+		return fallback
 	}
-
-	claimsBytes, err := json.Marshal(claims)
```

Comment:
If groups can have 2 types and Argo decide to support it. Simply validate the type in the jwt.MapClaims first, and if it is a string, set it to an array of 1 string before marshaling.